### PR TITLE
Core typed and documented

### PR DESCRIPTION
**Pull request by aidancrilly**
Addressing some issues: #29 #21 and #6 (for core.py at least, backend is not typed)

This is branched from PR #27 so this should be merged first

### CODE REVIEW

**Comment by aidancrilly** [comment]
@shimwell Let me know if you are happy to this solution #29

**Comment by shimwell** [comment]
Huge step forwards on documentation here. Many thanks for the changing to also return the std_dev, Much appreciated.

I see a few lines changed that could have floats added as type hints and a ton of pep8 changes. I can do the floats and perhaps one day we can make a PR that runs the whole code base through [black](https://github.com/psf/black) or [ruff](https://docs.astral.sh/ruff/) which will tidy things up nicely

File: src/NeSST/core.py [hunk 1]
```diff
@@ -34,27 +28,82 @@
...
+    Returns:
+        numpy.array : array with normalised TT spectral shape at energies E
+    
+    """
     return sm.TT_2dinterp(E,Tion)
 
-def yield_from_dt_yield_ratio(reaction,dt_yield,Tion,frac_D=frac_D_default,frac_T=frac_T_default):
-    ''' Reactivity ratio to predict yield from the DT yield assuming same volume and burn time
+def yield_from_dt_yield_ratio(reaction : str, dt_yield : float, Tion : float,
+                              frac_D=frac_D_default, frac_T=frac_T_default) -> float:
```
Comment (shimwell) [inline src/NeSST/core.py:82]:
```suggestion
                              frac_D: float = frac_D_default, frac_T: float = frac_T_default) -> float:
```

File: src/NeSST/core.py [hunk 2]
```diff
@@ -168,20 +218,39 @@ def DDprimspecmoments(Tion: float) -> typing.Tuple[float, float]:
...
+        Eout (numpy.array): the array on outgoing neutron energies
+    
+    """
     sm.mat_D.init_energy_grids(Eout,Ein)
     sm.mat_T.init_energy_grids(Eout,Ein)
     sm.mat_D.init_station_scatter_matrices()
     sm.mat_T.init_station_scatter_matrices()
 
-def init_DT_ionkin_scatter(varr,nT=False,nD=False):
+def init_DT_ionkin_scatter(varr : npt.NDArray, nT=False, nD=False):
```
Comment (shimwell) [inline src/NeSST/core.py:242]:
```suggestion
def init_DT_ionkin_scatter(varr : npt.NDArray, nT: bool = False, nD: bool = False):
```

File: src/NeSST/core.py [hunk 3]
```diff
@@ -193,7 +262,7 @@ def init_DT_ionkin_scatter(varr,nT=False,nD=False):
         else:
             sm.mat_D.full_scattering_matrix_create(varr)
 
-def calc_DT_ionkin_primspec_rhoL_integral(I_E,rhoL_func=None,nT=False,nD=False):
+def calc_DT_ionkin_primspec_rhoL_integral(I_E : npt.NDArray, rhoL_func=None, nT=False, nD=False):
```
Comment (shimwell) [inline src/NeSST/core.py:265]:
```suggestion
def calc_DT_ionkin_primspec_rhoL_integral(I_E : npt.NDArray, rhoL_func=None, nT: bool = False, nD: bool = False):
```

File: src/NeSST/core.py [hunk 4]
```diff
@@ -223,7 +303,26 @@ def init_mat_scatter(Eout,Ein,mat_label):
 #   Single Evalutation Scattered Spectra  #
 ###########################################
 
-def DT_sym_scatter_spec(I_E,frac_D=frac_D_default,frac_T=frac_T_default):
+def DT_sym_scatter_spec(I_E : npt.NDArray
+                        ,frac_D=frac_D_default, frac_T=frac_T_default
```
Comment (shimwell) [inline src/NeSST/core.py:307]:
```suggestion
                        ,frac_D: float = frac_D_default, frac_T: float = frac_T_default
```

File: src/NeSST/core.py [hunk 5]
```diff
@@ -234,7 +333,68 @@ def DT_sym_scatter_spec(I_E,frac_D=frac_D_default,frac_T=frac_T_default):
     total = nD+nT+Dn2n+Tn2n
     return total,(nD,nT,Dn2n,Tn2n)
 
-def DT_scatter_spec_w_ionkin(I_E,vbar,dv,rhoL_func,frac_D=frac_D_default,frac_T=frac_T_default):
+def DT_asym_scatter_spec(I_E : npt.NDArray, rhoL_func : callable,
+                         frac_D=frac_D_default, frac_T=frac_T_default
```
Comment (shimwell) [inline src/NeSST/core.py:337]:
```suggestion
                         frac_D: float = frac_D_default, frac_T: float = frac_T_default
```

File: src/NeSST/core.py [hunk 6]
```diff
@@ -234,7 +333,68 @@ def DT_sym_scatter_spec(I_E,frac_D=frac_D_default,frac_T=frac_T_default):
...
+    sm.mat_T.calc_dNdEs(I_E,rhoL_func)
+    nD   = frac_D*sm.mat_D.elastic_dNdE
+    nT   = frac_T*sm.mat_T.elastic_dNdE
+    Dn2n = frac_D*sm.mat_D.n2n_dNdE
+    Tn2n = frac_T*sm.mat_T.n2n_dNdE
+    total = nD+nT+Dn2n+Tn2n
+    return total,(nD,nT,Dn2n,Tn2n)
+
+def DT_scatter_spec_w_ionkin(I_E : npt.NDArray, vbar : float, dv : float, rhoL_func : callable,
+                             frac_D=frac_D_default, frac_T=frac_T_default
```
Comment (shimwell) [inline src/NeSST/core.py:371]:
```suggestion
                             frac_D: float = frac_D_default, frac_T: float = frac_T_default
```

File: src/NeSST/core.py (hunk 3 above)
Comment (shimwell) [reply src/NeSST/core.py:265]: perhaos rhoL_func should be a ```callable```, not sure

File: src/NeSST/core.py [hunk 7]
```diff
@@ -276,20 +447,44 @@ def mat_scatter_spec(mat,I_E,rhoL_func):
 # Full model fitting function #
 ###############################
 
-def calc_DT_sigmabar(Ein,I_E,frac_D=frac_D_default,frac_T=frac_T_default):
+def calc_DT_sigmabar(Ein : npt.NDArray, I_E : npt.NDArray,
+                     frac_D=frac_D_default, frac_T=frac_T_default) -> float:
```
Comment (shimwell) [inline src/NeSST/core.py:451]:
```suggestion
                     frac_D: float = frac_D_default, frac_T: float = frac_T_default) -> float:
```

File: src/NeSST/core.py [hunk 8]
```diff
@@ -276,20 +447,44 @@ def mat_scatter_spec(mat,I_E,rhoL_func):
...
+
+    Returns:
+        float : the spectrally averaged total DT cross section
+    """
     sigmabar = frac_D*np.trapz(sm.mat_D.sigma_tot(Ein)*I_E,Ein)+frac_T*np.trapz(sm.mat_T.sigma_tot(Ein)*I_E,Ein)
     return sigmabar
 
-def rhoR_2_A1s(rhoR,frac_D=frac_D_default,frac_T=frac_T_default):
+def rhoR_2_A1s(rhoR : typing.Union[float,npt.NDArray],
+               frac_D=frac_D_default,frac_T=frac_T_default) -> typing.Union[float,npt.NDArray]:
```
Comment (shimwell) [inline src/NeSST/core.py:465]:
```suggestion
               frac_D: float = frac_D_default,frac_T: float = frac_T_default) -> typing.Union[float,npt.NDArray]:
```

File: src/NeSST/core.py [hunk 9]
```diff
@@ -276,20 +447,44 @@ def mat_scatter_spec(mat,I_E,rhoL_func):
...
+    """
     mbar  = (frac_D*sm.A_D+frac_T*sm.A_T)*Mn_kg
-    # barns to m^2
-    sigmabarn = 1e-28
     A_1S = rhoR*(sigmabarn/mbar)
     return A_1S
 
-def A1s_2_rhoR(A_1S,frac_D=frac_D_default,frac_T=frac_T_default):
+def A1s_2_rhoR(A_1S : typing.Union[float,npt.NDArray],
+               frac_D=frac_D_default,frac_T=frac_T_default) -> typing.Union[float,npt.NDArray]:
```
Comment (shimwell) [inline src/NeSST/core.py:479]:
```suggestion
               frac_D: float = frac_D_default,frac_T: float = frac_T_default) -> typing.Union[float,npt.NDArray]:
```

File: src/NeSST/core.py [hunk 10]
```diff
@@ -234,7 +333,68 @@ def DT_sym_scatter_spec(I_E,frac_D=frac_D_default,frac_T=frac_T_default):
...
+    the incoming energy grid used to calculate the scattering matrices
+
+    The areal density function rhoL_func needs to be a callable function with a 
+    single argument (cosine[theta])
+
+    Args:
+        I_E (numpy.array): the neutron spectrum at Ein energies
+        rhoL_func (callable): must be a single argument function f(x), 
+        where x e [-1,1] and f(x) e [0,inf] and int f(x) dx = 1
+
```
Comment (shimwell) [inline src/NeSST/core.py:355]:
```suggestion
        frac_D (float) : fraction of D in fuel
        frac_T (float) : fraction of T in fuel

```

File: src/NeSST/core.py [hunk 11]
```diff
@@ -234,7 +333,68 @@ def DT_sym_scatter_spec(I_E,frac_D=frac_D_default,frac_T=frac_T_default):
...
+    The areal density function rhoL_func needs to be a callable function with a 
+    single argument (cosine[theta])
+
+    Args:
+        I_E (numpy.array): the neutron spectrum at Ein energies
+        vbar (float) : mean velocity of the scattering ions
+        dv (float) : standard deviation velocity of the scattering ions
+        rhoL_func (callable): must be a single argument function f(x), 
+        where x e [-1,1] and f(x) e [0,inf] and int f(x) dx = 1
+
```
Comment (shimwell) [inline src/NeSST/core.py:392]:
```suggestion
        frac_D (float) : fraction of D in fuel
        frac_T (float) : fraction of T in fuel

```

File: src/NeSST/core.py [hunk 12]
```diff
@@ -276,20 +447,44 @@ def mat_scatter_spec(mat,I_E,rhoL_func):
...
 
-def calc_DT_sigmabar(Ein,I_E,frac_D=frac_D_default,frac_T=frac_T_default):
+def calc_DT_sigmabar(Ein : npt.NDArray, I_E : npt.NDArray,
+                     frac_D=frac_D_default, frac_T=frac_T_default) -> float:
+    """Calculates the spectral-averaged cross section for DT
+
+    Args:
+        Ein (numpy.array): the array on incoming neutron energies
+        I_E (numpy.array): the neutron spectrum at Ein energies, assumed normalised
+
```
Comment (shimwell) [inline src/NeSST/core.py:457]:
```suggestion
        frac_D (float) : fraction of D in fuel
        frac_T (float) : fraction of T in fuel

```

File: src/NeSST/core.py [hunk 13]
```diff
@@ -276,20 +447,44 @@ def mat_scatter_spec(mat,I_E,rhoL_func):
...
     return sigmabar
 
-def rhoR_2_A1s(rhoR,frac_D=frac_D_default,frac_T=frac_T_default):
+def rhoR_2_A1s(rhoR : typing.Union[float,npt.NDArray],
+               frac_D=frac_D_default,frac_T=frac_T_default) -> typing.Union[float,npt.NDArray]:
+    """Calculates the scattering amplitude given a DT areal density in kg/m^2
+
+    Args:
+        rhoR (float): the DT areal density in kg/m^2
+
```
Comment (shimwell) [inline src/NeSST/core.py:470]:
```suggestion
        frac_D (float) : fraction of D in fuel
        frac_T (float) : fraction of T in fuel

```

File: src/NeSST/core.py [hunk 14]
```diff
@@ -276,20 +447,44 @@ def mat_scatter_spec(mat,I_E,rhoL_func):
...
     return A_1S
 
-def A1s_2_rhoR(A_1S,frac_D=frac_D_default,frac_T=frac_T_default):
+def A1s_2_rhoR(A_1S : typing.Union[float,npt.NDArray],
+               frac_D=frac_D_default,frac_T=frac_T_default) -> typing.Union[float,npt.NDArray]:
+    """Calculates the DT areal density in kg/m^2 given a scattering amplitude 
+
+    Args:
+        A_1S (float): the scattering amplitude for single scattering
+
```
Comment (shimwell) [inline src/NeSST/core.py:484]:
```suggestion
        frac_D (float) : fraction of D in fuel
        frac_T (float) : fraction of T in fuel

```

**Review by shimwell — APPROVED** [review]
I think this is a nice improvement and therefore certainly worth merging. I've added some optional changes but happy to see it merged with or without these changes

**Pull request merged**